people/models.py

```diff
@@ -0,0 +1,51 @@
+#  Gramps Online - an online version of the Gramps genealogy program
+#
+#  Copyright (C) 2018-2019 Brylie Christopher Oxley
+#
+#  This program is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU Affero General Public License as published
+#  by the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Affero General Public License for more details.
+#
+#  You should have received a copy of the GNU Affero General Public License
+#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from django.db import models
+
+class Person(models.Model):
+    """
+    A natural person,
+    modeled after GEDCOM X Person
+    http://www.gedcomx.org/v1/Person
+    """
+
+    # Choices for Person 'sex' field
```

Comment:
At present, Gramps only allows Male, Female and Unknown.  However, adding an extra option has been discussed in feature request [#5730](https://gramps-project.org/bugs/view.php?id=5730).  Intersex seems like a reasonable choice.

Reply:
Note that #5730 is about gender and this (as well as Gedcom X [issue 318](https://github.com/FamilySearch/gedcomx/issues/318) and [PR 320](https://github.com/FamilySearch/gedcomx/pull/320)) is about sex. This isn't the place for a discussion of how Gramps should model the two, but it's a discussion we need to have.

Reply:
@brylie Why not use the Gramps data model?  How much will the data models differ?

How difficult will it be to write and maintain the serializer/deserializer?  Will it be possible for Gramps users to test gramps-online with their own data before it is written?

Reply:
@brylie Have you considered using a MongoDB backend?  This would allow you to define the model directly on top of a JSON representation of the Gramps model.

Reply:
Well, I have been working with MongoDB for over three years now on a couple of other projects. Overall, I think it was a mistake, since our data have so many relationships. Since Gramps is built of relationships, a relational data model really makes sense.

Also, the PostgreSQL database has some really important plugins like PostGIS, which will be useful when modeling geospatial concepts.

Reply:
Our data is hierarchical rather than relational.

Anyway, continue for now.  I'm interested to see how this will develop.  I just thought that I would point out the first difference.

Reply:
Good point, thanks for the clarification.

Reply:
As an aside, I think that the hierarchical approach to genealogy may be akin to a 'flat-earth' worldview, in that more flix models like graphs would be more appropriate for the types of relationships that emerge in genealogical research (people relating with one another, places, events, etc. all sounds like a rich graph to me).

Reply:
> The idea is that a serializer/deserializer could be created in a separate repository (under MIT or some similar license), so the data could be portable between the two projects.

I think I would be opposed to *another* repository. But the code could go in the main gramps repository and would be license compatible with gramps-online.

The data model is a big topic and we should proceed with caution if we are considering modifying the existing model or creating a new one.

> Overall, I think it was a mistake

The current model is very Pythonic. It was chosen for expedience (keep in mind that Gramps was founded by *one* man). We should be careful about how we refer to the current model. It was designed/chosen in a very different time with very different context. I do not think it was a mistake. But as circumstances change, we  should consider changing Gramps to adapt.

> Anyway, continue for now. I'm interested to see how this will develop.

I agree!

Reply:
> I think I would be opposed to another repository.

For what it's worth, I was actually echoing the idea you mentioned during the license discussion, about having a repository with more permissive licensing.

> We should be careful about how we refer to the current model. It was designed/chosen in a very different time with very different context. I do not think it was a mistake.

The mistake I was referring to was using MongoDB for two projects I have helped to build ([Apinf](https://github.com/apinf/platform) and [GeriLife](https://github.com/GeriLife/wellbeing)).

Reply:
> Whereas connected models, like graphs, would be more appropriate for the types of relationships that emerge in genealogical research (people relating with one another, places, events, etc. all sounds like a rich graph to me).

Yes, parts of the Gramps model do resemble graphs.  At the same time I wrote the MongoDB backend, I also considered writing a neo4j backend.  Unfortunately, I couldn't come up with a workable design.

Another option that we may consider for Gramps in the future is a hybrid relational and json design.   Both SqLite and PostgreSql have good support for json.

One advantage of MongoDB is that we already have a backend.  No need to worry about data transfer.  We could even use the same database for the Gtk and web application.

I like the screenshots of GeriLife - very nice user interface!

Reply:
> Another option that we may consider for Gramps in the future is a hybrid relational and json design. Both SqLite and PostgreSql have good support for json.

That sounds like a good compromise. Although, I am working daily with a production PostgreSQL database where 30-50% of the fields are defined as JSONB. This has made some types of queries considerably harder than normalized queries. My take-away is that data should start off normalized, and then use JSONB where the sub-structure changes frequently or you need a specific performance optimization.

Reply:
For a new project with relational data I would certainly start with a fully normalised design.  I'm not sure that genealogical data is necessarily relational though.  This reminds me Tom Wetmore's phrase "the tyranny of relational thinking" I read in a post on the FHISO pubic mailing list.

Reply:
Heh, Tom's been banging that drum for at least 20 years, you can find it in his comments on the GENTECH data model in the old [GEDCOM-L mailing list](https://listserv.nodak.edu/cgi-bin/wa.exe?A0=GEDCOM-L) and in several of the Gedcom X discussions.